### PR TITLE
[WIP] Add add support for Timestamp type

### PR DESCRIPTION
**Pull request by frsyuki**
Implementation is still working in progress. 
Here is the spec of the new timestamp type: https://github.com/msgpack/msgpack/pull/209

### CODE REVIEW

File: msgpack-core/src/main/java/org/msgpack/core/MessagePacker.java [hunk 1]
```diff
@@ -32,6 +32,8 @@
 import java.nio.charset.CharsetEncoder;
 import java.nio.charset.CoderResult;
 import java.nio.charset.CodingErrorAction;
+import java.time.Instant;
```
Comment (xuwei-k) [inline msgpack-core/src/main/java/org/msgpack/core/MessagePacker.java]: drop Java 7 support ?

https://github.com/msgpack/msgpack-java/blob/81d540d55d86862c775f16b9457c586c9127a790/.travis.yml#L12-L13

Comment (xerial) [reply msgpack-core/src/main/java/org/msgpack/core/MessagePacker.java]: @frsyuki has confirmed it works in Java6 and Java7 even if Instant class is in the import statement. Only programs that uses Instance related methods will fail to compile.

**Comment by xerial** [comment]
TODO:
- [ ] Add timestamp MessageFormat types?

**Comment by mattbishop** [comment]
How is the timestamp feature coming along? I know the msgpack-js project is getting close to a release: https://github.com/kawanet/msgpack-lite/issues/76#issuecomment-347141538

**Comment by xerial** [comment]
@mattbishop We haven't decided how to add above two things; whether to add MessageFormat 
 type and ValueType for describing timestamp.

In the msgpack spec, timestamp is just one of the Extention types. We think adding these two additional types will be convenient for timestamp users. However, adding these types is a big change to the API (since most of the users is using switch case statement for checking data type, so rewriting will be necessary by adding `case Timestamp: ...`), so we need to be carful so as not to break user programs that started to use timestamp types in future versions.

File: msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java [hunk 2]
```diff
@@ -1232,6 +1246,70 @@ private String decodeStringFastPath(int length)
...
+                return (u32 & 0xffffffffL) * 1000L;
+            }
+            case 8: {
+                long data64 = readLong();
+                int nsec = (int) (data64 >>> 34);
+                long sec = data64 & 0x00000003ffffffffL;
+                return sec * 1000L + nsec / 1000L;
+            }
+            case 12: {
+                int nsec = readInt();
```
Comment (xerial) [inline msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java]: @frsyuki nsec is uint32, so &xffffffffL mask is necessary to convert negative int32 to uint32

File: msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java [hunk 3]
```diff
@@ -1232,6 +1246,70 @@ private String decodeStringFastPath(int length)
...
+    public Instant unpackInstant()
+            throws IOException
+    {
+        ExtensionTypeHeader ext = unpackExtensionTypeHeader();
+        if (ext.getType() != EXT_TIMESTAMP) {
+            throw unexpectedExtension("Timestamp", EXT_TIMESTAMP, ext.getType());
+        }
+        switch (ext.getLength()) {
+            case 4: {
+                int u32 = readInt();
```
Comment (xerial) [inline msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java]: @frsyuki Need to convert to uint32 by applying 0xFFFFFFFFL mask

File: msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java [hunk 4]
```diff
@@ -1232,6 +1246,70 @@ private String decodeStringFastPath(int length)
...
+                return Instant.ofEpochSecond(u32);
+            }
+            case 8: {
+                long data64 = readLong();
+                int nsec = (int) (data64 >>> 34);
+                long sec = data64 & 0x00000003ffffffffL;
+                return Instant.ofEpochSecond(sec, nsec);
+            }
+            case 12: {
+                int nsec = readInt();
```
Comment (xerial) [inline msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java]: This also needs to be converted to uint32 with 0xFFFFFFFFL mask

**Comment by gustf** [comment]
Hi there, any progress on this feature?

**Comment by xerial** [comment]
Continuing the work at #431

**Comment by xerial** [comment]
Closing this PR in favor of #565

**Pull request closed without merging**